src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java

```diff
@@ -15,26 +15,181 @@
+                    ? Long.MAX_VALUE : flowInfo.getMaxLatencyTier2();
+            fsms.put(getFsmKey(flowId, REVERSE), fsmFactory.produce(flowId, REVERSE, maxLatency, maxLatencyTier2));
+        }
+    }
 
     /**
      * Check flow SLA is violated.
      */
-    public void checkFlowSla(String flowId, FlowDirection direction, long latency,
-                             Long maxLatency, Long maxLatencyTier2) {
-        if (maxLatency != null && maxLatency > 0 && latency > maxLatency) {
-            long latencyMs = TimeUnit.NANOSECONDS.toMillis(latency);
-            long maxLatencyMs = TimeUnit.NANOSECONDS.toMillis(maxLatency);
-            log.warn("{} path for flow '{}' violates latency SLA {}/{} ms",
-                    direction.name(), flowId, latencyMs, maxLatencyMs);
-        }
-        //TODO add flow SLA violation reactions
+    public void processFlowLatencyMeasurement(String flowId, FlowDirection direction, long latency) {
+        String key = getFsmKey(flowId, direction.name().toLowerCase());
+        FlowLatencyMonitoringFsm fsm = fsms.get(key);
+        if (fsm == null) {
+            Flow flow = flowRepository.findById(flowId)
+                    .orElseThrow(() -> new IllegalStateException(format("Flow %s not found.", flowId)));
+            long maxLatency = flow.getMaxLatency() == null || flow.getMaxLatency() == 0
+                    ? Long.MAX_VALUE : flow.getMaxLatency();
+            long maxLatencyTier2 = flow.getMaxLatencyTier2() == null || flow.getMaxLatencyTier2() == 0
+                    ? Long.MAX_VALUE : flow.getMaxLatencyTier2();
+            fsm = fsmFactory.produce(flowId, direction.name().toLowerCase(), maxLatency, maxLatencyTier2);
+            fsms.put(key, fsm);
+        }
+
+        Event event = getCorrespondingEvent(fsm, latency);
+        Context context = Context.builder()
+                .latency(latency)
+                .carrier(this)
+                .build();
+        fsmExecutor.fire(fsm, event, context);
+    }
+
+    /**
+     * Process tick.
+     */
+    public void processTick() {
+        Context context = Context.builder()
+                .carrier(this)
+                .build();
+        fsms.values().forEach(fsm -> fsmExecutor.fire(fsm, Event.TICK, context));
+    }
+
+    private String getFsmKey(String flowId, String direction) {
+        return format("%s_%s", flowId, direction);
+    }
+
+    private Event getCorrespondingEvent(FlowLatencyMonitoringFsm fsm, long latency) {
+        long maxLatency = fsm.getMaxLatency();
+        long maxLatencyTier2 = fsm.getMaxLatencyTier2();
+        switch (fsm.getLastStableState()) {
+            case HEALTHY:
+                maxLatency = (long) (maxLatency * (1 + threshold));
+                break;
+            case TIER_1_FAILED:
+                maxLatency = (long) (maxLatency * (1 - threshold));
+                maxLatencyTier2 = (long) (maxLatencyTier2 * (1 + threshold));
```

Comment:
In line 107 and 110 maxLatency field can be set to Long.MAX_VALUE. And here you multiply in on 1.05. Looks like overflow

Reply:
Fixed.

Reply:
As it turned out special case for Long.MAX_VALUE is not required here. During float to long cast Java takes care of overflows. Special case treatment removed.